### PR TITLE
DM-1682 [App] Improve Search Speed from Homepage

### PR DESCRIPTION
**Pull request by ahay-agile6**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-1682

## Description - what does this code do?
Caches the searchable practices query result for use in areas of the application

## Testing done - how did you test it/steps on how can another person can test it 
- TURN ON REDIS!
- Make sure it is running
- Have a terminal running with a rails console running
- Be sure to `bundle install` and restart the server
- have a stopwatch handy
- have lots of practices in the db
- also, do not hit the homepage before the test, if you do, use the rails console to clear the cache: ```Rails.cache.delete('searchable_practices')``` this is because the homepage and the search page use the same query to get the list of practices

1. browse to `/search`, use the stopwatch to see how long the page takes to load
2. browse back to the home page, or any other page that is not search
3. (Optional) in the open rails console, check if rails cached the practices in redis: ```Rails.cache.fetch('searchable_practices')``` you should get a giant list of practices
4. browse to `/search` again, use the stopwatch to see how long the page takes to load
5. is it faster? might be the same locally, i dunno
6. update a practice's tagline
7. browse to `/search` again
8. it should be the same speed
9. you can check if the cache was updated in the rails console: 
```Rails.cache.fetch('searchable_practices').find(<id of practice you updated>).tagline```

## Acceptance criteria
- [ ]

## Definition of done
- [x] Unit tests written (if applicable)
- [x] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating JIRA issue
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -63,8 +63,9 @@ def index
   end
 
   def search
+    # TODO: why do I have this here? - A.H.
```
Comment (ahay-agile6) [inline app/controllers/home_controller.rb:66]: i'll figure this out eventually

File: app/models/practice.rb [hunk 2]
```diff
@@ -17,6 +21,30 @@ class Practice < ApplicationRecord
...
+        self.description_changed? ||
+        self.summary_changed? ||
+        self.initiating_facility_changed? ||
+        self.main_display_image_updated_at_changed? ||
+        self.published_changed?
+      clear_cache
+    end
+  end
+
+  def self.searchable_practices
```
Comment (ahay-agile6) [inline app/models/practice.rb:42]: this is the reusable method that either stores or fetches from the cache.
if the cache from redis is nil, it runs the query in the block and caches it.
neato, huh?

File: spec/features/search_spec.rb [hunk 3]
```diff
@@ -36,23 +82,57 @@
...
+      click_link('New Practice')
+      fill_in('Practice name', with: 'The Newest Practice')
+      fill_in('User email', with: 'practice_owner@va.gov')
+      click_button('Create Practice')
+      latest_practice = Practice.last
+
+      visit '/search?=newest'
+      expect(page).to_not have_content(latest_practice.name)
+
+      publish_practice(latest_practice)
```
Comment (emyl3) [inline spec/features/search_spec.rb:129]: the tests kept failing both locally and in CI after the publish... 
`find('h1#overview', text: latest_practice.name)`
adding this hack-y wait helped the tests to pass... 😅 let me know if you have a more elegant alternative 😄

Comment (ahay-agile6) [reply spec/features/search_spec.rb:129]: thanks! let's hack away :axe:

**Comment by ahay-agile6** [comment]
this was mostly @JoshingYou1  :slightly_smiling_face:

**Comment by emyl3** [comment]
Good work @JoshingYou1 😹 🎉 🤩

**Pull request merged**